Index: b/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
===================================================================
--- a/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
@@ -162,6 +162,7 @@
 	rx_delay = <0x18>;
 	snps,aal;
 	tx_delay = <0x24>;
+	handle_cpu_id = <1>;
 	status = "okay";
 
 	mdio {
@@ -183,10 +184,16 @@
 };
 
 &i2c0 {
+	clock-frequency = <200000>;
+	i2c-scl-rising-time-ns = <150>;
+	i2c-scl-falling-time-ns = <30>;
 	status = "okay";
 };
 
 &i2c1 {
+	clock-frequency = <400000>;
+	i2c-scl-rising-time-ns = <160>;
+	i2c-scl-falling-time-ns = <30>;
 	status = "okay";
 
 	rk805: pmic@18 {
@@ -369,6 +376,7 @@
 
 &sdmmc {
 	bus-width = <4>;
+	cap-mmc-highspeed;
 	cap-sd-highspeed;
 	disable-wp;
 	pinctrl-0 = <&sdmmc0_clk>, <&sdmmc0_cmd>, <&sdmmc0_dectn>, <&sdmmc0_bus4>;
@@ -412,6 +420,7 @@
 &usbdrd3 {
 	dr_mode = "host";
 	status = "okay";
+	handle_cpu_id = <2>;
 	#address-cells = <1>;
 	#size-cells = <0>;
 
@@ -419,7 +428,7 @@
 	rtl8153: usb-eth@2 {
 		compatible = "realtek,rtl8153";
 		reg = <2>;
-
+		local-mac-address = [00 00 00 00 00 00];
 		realtek,led-data = <0x87>;
 	};
 };
Index: b/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus.dts
===================================================================
--- a/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus.dts
@@ -129,6 +129,7 @@
 	snps,aal;
 	rx_delay = <0x18>;
 	tx_delay = <0x24>;
+	handle_cpu_id = <1>;
 	status = "okay";
 
 	mdio {
@@ -148,6 +149,9 @@
 };
 
 &i2c1 {
+	clock-frequency = <400000>;
+	i2c-scl-rising-time-ns = <160>;
+	i2c-scl-falling-time-ns = <30>;
 	status = "okay";
 
 	rk805: pmic@18 {
@@ -318,6 +322,7 @@
 
 &sdmmc {
 	bus-width = <4>;
+	cap-mmc-highspeed;
 	cap-sd-highspeed;
 	disable-wp;
 	pinctrl-0 = <&sdmmc0_clk>, <&sdmmc0_cmd>, <&sdmmc0_dectn>, <&sdmmc0_bus4>;
@@ -327,7 +332,7 @@
 };
 
 &spi0 {
-	status = "okay";
+	status = "disabled";
 
 	flash@0 {
 		compatible = "jedec,spi-nor";
@@ -366,6 +371,7 @@
 &usbdrd3 {
 	dr_mode = "host";
 	status = "okay";
+	handle_cpu_id = <2>;
 	#address-cells = <1>;
 	#size-cells = <0>;
 
@@ -373,7 +379,7 @@
 	rtl8153: usb-eth@2 {
 		compatible = "realtek,rtl8153";
 		reg = <2>;
-
+		local-mac-address = [00 00 00 00 00 00];
 		realtek,led-data = <0x87>;
 	};
 };
@@ -385,3 +391,16 @@
 &usb_host0_ohci {
 	status = "okay";
 };
Index: b/arch/arm64/boot/dts/rockchip/rk3328.dtsi
===================================================================
--- a/arch/arm64/boot/dts/rockchip/rk3328.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3328.dtsi
@@ -148,12 +148,7 @@
 		};
 		opp-1512000000 {
 			opp-hz = /bits/ 64 <1512000000>;
-			opp-microvolt = <1450000>;
-			clock-latency-ns = <40000>;
-		};
-		opp-1608000000 {
-			opp-hz = /bits/ 64 <1608000000>;
-			opp-microvolt = <1450000>;
+			opp-microvolt = <1420000>;
 			clock-latency-ns = <40000>;
 		};
 	};
Index: b/drivers/net/phy/motorcomm.c
===================================================================
--- a/drivers/net/phy/motorcomm.c
+++ b/drivers/net/phy/motorcomm.c
@@ -1525,6 +1525,13 @@ static int yt8531_config_init(struct phy
 			return ret;
 	}
 
+	/* LED0: Unused/Off, LED1: Link, LED2: Activity, 8Hz */
+	ytphy_write_ext(phydev, 0xa00b, 0xe004);
+	ytphy_write_ext(phydev, 0xa00c, 0);
+	ytphy_write_ext(phydev, 0xa00d, 0x2600);
+	ytphy_write_ext(phydev, 0xa00e, 0x0070);
+	ytphy_write_ext(phydev, 0xa00f, 0x000a);
+
 	return 0;
 }
 
 