Index: b/arch/arm/dts/rk3328-orangepi-r1-plus-lts-u-boot.dtsi
===================================================================
--- a/arch/arm/dts/rk3328-orangepi-r1-plus-lts-u-boot.dtsi
+++ b/arch/arm/dts/rk3328-orangepi-r1-plus-lts-u-boot.dtsi
@@ -1 +1,34 @@
-#include "rk3328-orangepi-r1-plus-u-boot.dtsi"
+#include "rk3328-u-boot.dtsi"
+#include "rk3328-sdram-lpddr3-666.dtsi"
+/ {
+	chosen {
+		u-boot,spl-boot-order = "same-as-spl", &sdmmc, &emmc;
+	};
+};
+
+&gpio0 {
+	u-boot,dm-spl;
+};
+
+&pinctrl {
+	u-boot,dm-spl;
+};
+
+&sdmmc0m1_gpio {
+	u-boot,dm-spl;
+};
+
+&pcfg_pull_up_4ma {
+	u-boot,dm-spl;
+};
+
+/* Need this and all the pinctrl/gpio stuff above to set pinmux */
+&vcc_sd {
+	u-boot,dm-spl;
+};
+
+&gmac2io {
+	snps,reset-gpio = <&gpio1 RK_PC2 GPIO_ACTIVE_LOW>;
+	snps,reset-active-low;
+	snps,reset-delays-us = <0 15000 50000>;
+};
Index: b/arch/arm/dts/rk3328-orangepi-r1-plus.dts
===================================================================
--- a/arch/arm/dts/rk3328-orangepi-r1-plus.dts
+++ b/arch/arm/dts/rk3328-orangepi-r1-plus.dts
@@ -11,7 +11,7 @@
 };
 
 &spi0 {
-	status = "okay";
+	status = "disabled";
 
 	flash@0 {
 		compatible = "jedec,spi-nor";
Index: b/board/rockchip/evb_rk3328/MAINTAINERS
===================================================================
--- a/board/rockchip/evb_rk3328/MAINTAINERS
+++ b/board/rockchip/evb_rk3328/MAINTAINERS
@@ -19,6 +19,13 @@ F:      configs/orangepi-r1-plus-rk3328_
 F:      arch/arm/dts/rk3328-orangepi-r1-plus-u-boot.dtsi
 F:      arch/arm/dts/rk3328-orangepi-r1-plus.dts
 
+ORANGEPI-R1-PLUS-LTS-RK3328
+M:      Shenzhen Xunlong Software CO.,Limited <zhao_steven@263.net>
+S:      Maintained
+F:      configs/orangepi-r1-plus-lts-rk3328_defconfig
+F:      arch/arm/dts/rk3328-orangepi-r1-plus-lts-u-boot.dtsi
+F:      arch/arm/dts/rk3328-orangepi-r1-plus-lts.dts
+
 ROC-RK3328-CC
 M:      Loic Devulder <ldevulder@suse.com>
 M:      Chen-Yu Tsai <wens@csie.org>
