Index: b/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
===================================================================
--- a/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
@@ -216,6 +216,7 @@
 	rx_delay = <0x18>;
 	snps,aal;
 	tx_delay = <0x24>;
+	handle_cpu_id = <1>;
 	status = "okay";
 
 	mdio {
@@ -237,10 +238,16 @@
 };
 
 &i2c0 {
+	clock-frequency = <200000>;
+	i2c-scl-rising-time-ns = <150>;
+	i2c-scl-falling-time-ns = <30>;
 	status = "okay";
 };
 
 &i2c1 {
+	clock-frequency = <400000>;
+	i2c-scl-rising-time-ns = <160>;
+	i2c-scl-falling-time-ns = <30>;
 	status = "okay";
 
 	rk805: pmic@18 {
@@ -423,8 +430,10 @@
 
 &sdmmc {
 	bus-width = <4>;
+	cap-mmc-highspeed;
 	cap-sd-highspeed;
 	disable-wp;
+	max-frequency = <150000000>;
 	pinctrl-0 = <&sdmmc0_clk>, <&sdmmc0_cmd>, <&sdmmc0_dectn>, <&sdmmc0_bus4>;
 	pinctrl-names = "default";
 	sd-uhs-sdr12;
@@ -478,11 +487,30 @@
 &usbdrd_dwc3 {
 	dr_mode = "host";
 	status = "okay";
+	handle_cpu_id = <2>;
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	snps,xhci-slow-suspend-quirk;
+	snps,xhci-trb-ent-quirk;
 
 	usb_eth: usb-eth@2 {
 		compatible = "realtek,rtl8153";
 		reg = <2>;
-
+		local-mac-address = [00 00 00 00 00 00];
 		realtek,led-data = <0x87>;
 	};
 };
+
+&u3phy {
+	vbus-supply = <&vdd_5v>;
+	status = "okay";
+};
+
+&u3phy_utmi {
+	status = "okay";
+};
+
+&u3phy_pipe {
+	status = "okay";
+};
Index: b/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus.dts
===================================================================
--- a/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus.dts
@@ -12,7 +12,7 @@
 
 &spi0 {
 	max-freq = <48000000>;
-	status = "okay";
+	status = "disabled";
 
 	flash@0 {
 		compatible = "jedec,spi-nor";
Index: b/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus-lts.dts
===================================================================
--- a/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus-lts.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus-lts.dts
@@ -14,13 +14,13 @@
 };
 
 &dmc_opp_table {
-	opp-798000000 {
+	opp-786000000 {
 		status = "disabled";
 	};
-	opp-840000000 {
+	opp-798000000 {
 		status = "disabled";
 	};
-	opp-924000000 {
+	opp-840000000 {
 		status = "disabled";
 	};
 	opp-1056000000 {
@@ -44,6 +44,7 @@
 			pinctrl-names = "default";
 			reset-assert-us = <15000>;
 			reset-deassert-us = <50000>;
+			keep-clkout-on;
 			reset-gpios = <&gpio1 RK_PC2 GPIO_ACTIVE_LOW>;
 		};
 	};
